feat: Revenue Analytics v0

rafaeelaudibert opened the pull request:
## Changes

First attempt at creating a virtual view for our revenue analytics purposes. This should abstract away the different revenue providers we'll have into a shared view used by different providers. It only supports Stripe now but it's easily extensible.

I've also removed all of the old code that made this extremely generic. We don't need that because we believe providing the user with a good default is the better UX here.

This will eventually become a materialized view, but for now it's just an easy way to extract and manipulate the data in a single place rather than copying it around in different places - and making generating insights extremely complicated.

To do:
- expose this view to Max AI
- expose this view in the Database instrospection
- get tests to pass, need @PostHog/team-data-warehouse help here

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

Manual + unit tests (still need to pass)
Also, this is all hidden behind a new FF

rafaeelaudibert committed:
feat: Add basic data warehouse structure

This should let us test the new data warehouse structure for revenue
analytics in our test page (inside data management).

We'll not be adding these to web analytics because it's just too
tricky: getting access to that data *from* the frontend is tough, we'll
need new query runners to make this work well.

We can still relay to `InsightsQuery`s but it will need to be done in
the backend - like Web Vitals, for example.

rafaeelaudibert commented on products/revenue_analytics/backend/models.py:
```diff
@@ -0,0 +1,203 @@
+    CurrencyCode.XPF,
+]
+
+BASE_FIELDS: dict[str, FieldOrTable] = {
```
This + some lines below is where we're creating the view. This will eventually become a materialized view that's automatically created by the system, but this can live like a virtual view for now.

rafaeelaudibert commented on products/revenue_analytics/manifest.tsx:
```diff
@@ -0,0 +1,38 @@
+    urls: {
+        revenueAnalytics: (): string => '/revenue_analytics',
+    },
+    fileSystemTypes: {
+        // TODO: add to project tree backend
+        // revenue_analytics: {
+        //     icon: <IconRocket />,
+        //     href: (ref: string) => urls.revenueAnalytics(ref),
+        // },
+    },
```
I assume I dont need anything here, right @mariusandra

mariusandra replied:
You can leave this blank as I don't think you're adding new file types, but `treeItems` below should be `treeItemsNew` (adds to the context menu as "new X") or `treeItemsExplore` (not used now, but we'll replace the sidebar with these one day).

rafaeelaudibert replied:
Nice, thanks!

Suggestion: a nested `treeItems: { new: [], explore: [] }` sounds nicer :)

mariusandra replied:
It definitely does, and we might need even more. Luckily this is all easy to refactor once we actually start using this data in the sidebar.

rafaeelaudibert commented on products/revenue_analytics/backend/hogql_queries/test/test_revenue_example_data_warehouse_tables_query_runner.py:
```diff
@@ -0,0 +1,107 @@
+
+        assert len(results) == 0
+
+    # TODO: These fail because it'll complain the table doesn't exist
+    # posthog.errors.CHQueryErrorUnknownTable: Code: 60.
+    # DB::Exception: Unknown table expression identifier 'stripe_charge'
+    @patch("posthoganalytics.feature_enabled", return_value=False)
+    def test_database_query(self, feature_enabled_mock):
+        response = self._run_revenue_example_external_tables_query()
+        results = response.results
+
+        assert len(results) == 3
+
+    # TODO: These fail because it'll complain the table doesn't exist
+    # posthog.errors.CHQueryErrorUnknownTable: Code: 60.
+    # DB::Exception: Unknown table expression identifier 'stripe_charge'
+    @patch("posthoganalytics.feature_enabled", return_value=True)
+    def test_database_query_with_currency_conversion(self, feature_enabled_mock):
+        response = self._run_revenue_example_external_tables_query()
+        results = response.results
+
+        assert len(results) == 3
```
@PostHog/team-data-warehouse I need some help here. I want to be able to run these locally, but it won't run because:
- Im using a parquet file that doesn't exist, how do I make it exist? Can I make it a CSV instead? I see team experiments put some effort into this, but it's a lot of work, and it requires some Docker internal host handling that I'm not knowledgeable about
- How do I make Clickhouse know about these tables/views? Do I need to create them somehow?

rafaeelaudibert replied:
I'll skip these for now in the interest of time and will come back in the future. I'll add a `skip` annotation to these tests

rafaeelaudibert replied:
Im now working on getting these to pass using a local CSV while I don't get an approve from @team-dawa

EDsCODE replied:
Context here: https://posthog.slack.com/archives/C019RAX2XBN/p1743724035504849

robbie-c commented on posthog/hogql/database/database.py:
```diff
@@ -261,11 +263,14 @@ def create_hogql_database(
 
-    if timings is None:
-        timings = HogQLTimings()
+    if not team_arg and not team_id:
```
Can team_id be 0?

```suggestion
    if not team_arg and team_id is None:
```

rafaeelaudibert replied:
yeah, makes sense, fixing it

robbie-c commented on posthog/hogql_queries/query_runner.py:
```diff
@@ -377,7 +377,9 @@ def get_query_runner(
 
     if kind == "RevenueExampleEventsQuery":
-        from .web_analytics.revenue_example_events_query_runner import RevenueExampleEventsQueryRunner
+        from products.revenue_analytics.backend.hogql_queries.revenue_example_events_query_runner import (
```
🎉

robbie-c commented:
I'm happy with the web analytics part - will let the dw review the dw stuff :)

rafaeelaudibert commented on products/conftest.py:
```diff
@@ -0,0 +1,209 @@
+# NOTE: This file is hardlinked both on posthog/conftest.py and products/conftest.py
```
Hm, I thought git would keep the hardlinks :/

Symlinks aren't working, pytest can't find them. I'll simply duplicate the file and note that they should be kept in sync

rafaeelaudibert replied:
Didnt duplicate, just imported from the original posthog/conftest and its working great!

rafaeelaudibert commented on products/revenue_analytics/backend/hogql_queries/revenue_example_data_warehouse_tables_query_runner.py:
```diff
@@ -0,0 +1,91 @@
+        for view_name in self.hogql_context.database.get_views():
+            view = self.hogql_context.database.get_table(view_name)
+            if isinstance(view, RevenueAnalyticsRevenueView):
+                view = cast(RevenueAnalyticsRevenueView, view)
```
I do, pytest is much much more naive than Typescript :(

rafaeelaudibert committed:
fix: Impor conftest data

Rather than using a hardlink like I did before - doesnt interact well with git - lets simply import everything, effectively copying all of the declarations

rafaeelaudibert committed:
refactor: Simplify `create_hogql_database` calls

Lets now make sure we only need to pass either the `team_id` OR `team` as a parameter

rafaeelaudibert committed:
fix: Fix some mypy errors

rafaeelaudibert committed:
fix: Include new product in base package.json

rafaeelaudibert committed:
fix: Remove references to data warehouse in revenue logic

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

github-actions[bot] committed:
Update query snapshots

rafaeelaudibert committed:
refacotr: Simplify query setup

We dont need to pass these in, we have access to it in the backend already

rafaeelaudibert committed:
fix: Fix manifest

Manifest was slightly outdated - team devex has been working on this structure

rafaeelaudibert committed:
ci: Skip tests while were working on them

EDsCODE commented on posthog/temporal/batch_exports/spmc.py:
```diff
@@ -1090,7 +1090,7 @@ def compose_filters_clause(
         modifiers=HogQLQueryModifiers(materializationMode=MaterializationMode.DISABLED),
     )
-    context.database = create_hogql_database(team.id, context.modifiers)
+    context.database = create_hogql_database(team=team, modifiers=context.modifiers)
```
Highly suggest making this param change in a separate PR to make this one more reviewable

rafaeelaudibert replied:
I agree. I've decided to split this PR into several PRs, this will make reviewing (and merging) this much easier.

EDsCODE commented on products/revenue_analytics/backend/models.py:
```diff
@@ -0,0 +1,222 @@
+}
+
+
+class RevenueAnalyticsRevenueView(Table):
```
Is this necessary to create another type? I'm wondering if we could just have these fields written in as a "SavedQuery" https://github.com/PostHog/posthog/blob/22504acef2da55e8d8bc4d2648c002b321166226/posthog/hogql/database/models.py#L231

rafaeelaudibert replied:
Yeah, reusing this was much easier, thanks for the suggestion!

rafaeelaudibert committed:
feat: Add tests to data warehouse

Finally add tests to the data warehouse, no mocking whatsoever, this is a very elegant fix, I'll have a separate PR copying this implementation to other places

rafaeelaudibert committed:
tests: Fix query count

rafaeelaudibert commented:
I've decided to break this PR into several smaller PRs to make this more readable and reviewable. I'll have all of them open next Monday, keep an eye out :)

Specially @EDsCODE and @robbie-c @lricoy 

PRs:

- Remove currency conversion FF
- Refactor `create_hogql_database`
- Introduce new helper function for Data Warehouse setup via mock CSV files
- Actual Revenue Analytics code/tables

github-actions[bot] committed:
Update query snapshots

rafaeelaudibert commented:
Superseded by https://github.com/PostHog/posthog/pull/30895
